src/compiler/checker.ts

```diff
@@ -6114,6 +6131,15 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
                 approximateLength: 0,
                 trackedSymbols: undefined,
                 bundled: !!compilerOptions.outFile && !!enclosingDeclaration && isExternalOrCommonJsModule(getSourceFileOfNode(enclosingDeclaration)),
+                isEntityNameVisible: (entityName, shouldComputeAliasToMakeVisible) => isEntityNameVisible(entityName, context.enclosingDeclaration!, shouldComputeAliasToMakeVisible),
+                isExpandoFunctionDeclaration,
+                isNonNarrowedBindableName,
+                getAllAccessorDeclarations: getAllAccessorDeclarationsForDeclaration,
+                requiresAddingImplicitUndefined,
+                isUndefinedIdentifierExpression(node: Identifier) {
+                    Debug.assert(isExpressionNode(node));
+                    return getSymbolAtLocation(node) === undefinedSymbol;
+                },

```

Comment:
Every time we clone the context, we're copying these functions over and over again even though they never change. I'm guessing this is why memory usage in the benchmarks has increased. We should split these off onto another object. Probably even, construct the object (a host, I guess?) at the top of the checker.

This refactor also exposes what looks like a bug; `isEntityNameVisible` closes over `context`, but `context` can be cloned but `isEntityNameVisible` will still refer to the old object. We screw around with `context.enclosingDeclaration` all over the place, so this could potentially lead to a bug.

Reply:
I was planning on doing exactly this in my follow on PRs. I can try to move it now if this is critical before the beta.

Reply:
I have a change locally I could push up, nearly finished.

Reply:
That would be great. Thank you.